clusters/build-clusters/01_cluster/Makefile

```diff
@@ -0,0 +1,57 @@
+release_image := registry.svc.ci.openshift.org/ocp/release:4.2.0-0.nightly-2019-10-08-232417
+
+.PHONY: all
+all:
+	#TODO (hongkliu)
+	@echo "target 'all' has not yet implemented"
+
+.PHONY: install-dptp-managed-cluster
+install-dptp-managed-cluster:
+	@echo "installing dptp-managed cluster ..."
+	./install_cluster.sh $(release_image)
+
+###pip install --user yq
+secret_file := "/tmp/secret.yaml"
+client_id := $(shell yq -r .github.client_id $(secret_file))
+client_secret := $(shell yq -r .github.client_secret $(secret_file))
+
+.PHONY: install-configure-github-oauth
+install-configure-github-oauth:
```

Comment:
Can we *please* avoid constructing these elaborate `Makefile` targets as much as possible, and use `applyconfig` instead? We tried to go from `Makefile` targets for a reason - it's hard to dry-run changes, you need to maintain Makefiles and also need to not forget to hook new resources into correct targets... We've been there. 

Can you instead create this so that, provided I have a kubeconfig pointing to an empty cluster, `applyconfig` does all this? Now, `applyconfig` probably does not have all features you need, like filing in secrets, but they should not be that hard to implement. It will save us a lot of long-term pain.

Reply:
So I can assume the sensitive information that I need for those objects is saved in some secret in the cluster ... and `applyConfig` can get the entry from the secret and apply it into our objects? This will be another Jira card for applyConfig I think, right?

Which component will maintain the secrets then? I think the manual work like we do via `bitwarden.script`?

Reply:
Yeah, I think that would be a reasonable way to proceed: have secrets in the cluster and make `applyconfig` interact with them, possibly using templates (which applyconfig can already do, just cannot set template parameters yet). It will become a little tricky timing-wise, because to sync secrets, you need namespaces, but to create namespaces, you need to run `applyconfig`... ugh.

Tricky but solvable, IMHO (and worth solving).

Reply:
Great. I will create a JIRA card for `applyConfig`.

Reply:
The os-template supports only the value for vars as a constant ... we need something like

```
parameters:
- description: github_client_id
  displayName: github_client_id
  name: github_client_id
  required: true
  value: "$(oc get secret -n <ns> <secret_name> -o jsonpath ... | base64 -d)"
```

when `applyConfig` [detects a template](https://github.com/openshift/ci-tools/blob/master/cmd/applyconfig/applyconfig.go#L160), it does `cmd` to vendor the value.

Does it sound a good plan?

Reply:
I would prefer to avoid building the ability to execute arbitrary code into applyconfig, if possible.

I think we need this feature mostly (only?) for secrets - everything else can be committed to the repository. Hence, I think we can do something more targeted to secrets: specifying a `secret.item`, then applyconfig would fetch this value and fill it it.

Or, probably even better, we could make `applyconfig` stupid (I like the idea of it being stupid to avoid writing code interfacing with secrets) and just passthrough environmental variables to `oc process` - if `$github_client_id` is defined, pass it as a parameter. Caller would then be responsible for preparing these environmental variables. We would simply expose these secrets as environmental variables in the jobs running `applyconfig`.

Reply:
> Hence, I think we can do something more targeted to secrets: specifying a secret.item, then applyconfig would fetch this value and fill it it.

One more question: If a non-secret object wants to use a field in secret, what is the logic?
For example, 

https://github.com/openshift/release/blob/60cfe5965492d743fae069702a04af68d1bc1036/clusters/build-clusters/01_cluster/assets/cluster_oauth.yaml#L2-L12

Reply:
> If a non-secret object wants to use a field in secret

I'm not sure I understand this correctly, do you mean that the `clientID` value is a secret too, so it can't be committed to the repo? I think this is exactly the use case for templates: you would pass that as a parameter.

Reply:
Yes, client ID is just an example of some field is from a secret too. How can `applyConfig` handle this case?

This is from the [github doc](https://developer.github.com/v3/guides/basics-of-authentication/)

_Your client ID and client secret keys come from your application's configuration page. You should never, ever store these values in GitHub_

When I talked about oauth setting with the cblecker, I got the impression that client id is not that sentitive ... maybe that is why `OAuth` CRD does not support getting it from a secret? Not sure.

Reply:
I think I need a bj session with you. I am confused by my OWN target here.

For secrets, the plan is `secret-mirroring` will support to mirror secret cross clusters (how the secret is generated is out of scope of this PR. I would assume to use the bitwarden script).

We use a makefile target here to generate the secret is a workaround before the above is implemented. They will be removed afterwards.

So we can assume that all the secrets that we need are there in the build cluster.

For a CRD which needs a field in secret:

https://github.com/openshift/release/blob/60cfe5965492d743fae069702a04af68d1bc1036/clusters/build-clusters/01_cluster/assets/cluster_oauth.yaml#L1-L16

If the CRD supports to read the value from a secret (e.g., `clientSecret` above), we do not need to do anything there.

If the CRD does NOT support to read the value from a secret (e.g., `clientID` above), and we do not want to commit its value into git-repo, we need a way to give its value. I'd like to discuss with you about the best way to do it. I think we could use the template (in this case, the item is `OAuth` instead of `secret` in the template)

```
parameters:
- name: github_client_id
  required: false
  value: "${github_client_id}"
```

When `applyConfig` sees `${github_client_id}` in the `value` field, it will os.getEnv("github_client_id") to vendor its value.

Reply:
> When applyConfig sees ${github_client_id} in the value field, it will os.getEnv("github_client_id") to vendor its value.

I don't think we even need special value fields. The parameter has a name and we can use that name also as a name of the expected environmental variable. So, when processing a template, applyconfig can for each parameter check whether a envvar is defined and if so, supply its value as a parameter value for the template.

Reply:
Also, an observation: in this case, the important difference between the two secrets is where they need to be present. The `clientSecret` that the `OAuth` CR refers to must be present on the build cluster, while the `clientId` secret (that we want to fill into the CR as a value using templating), needs to be available to the place running applyconfig, which in this specific case is probably a developer machine, as our automation does not handle cluster-level config now. If it did, though, the pod running the setup for new build cluster would probably not be running on this new build cluster, but somewhere else (where? another build cluster? (what would build the first one?) CI cluster?).

Not something we need to act on right now, but good to have in mind.

Reply:
Yes, now the picture is clearer too me.
Let me draft a config/secret design for review, then we will have some common grounds we can follow.

Reply:
> The parameter has a name and we can use that name also as a name of the expected environmental variable.

The would restrict the naming of the params, e.g., `HOME` and maybe the value of $HOME is a different story.

Reply:
> The would restrict the naming of the params, e.g., HOME and maybe the value of $HOME is a different story.

I know, I think it's acceptable.

Reply:
ACK.